File: packages/shared/src/fidelidade.ts
// Fidelidade: a cada 10 pizzas G -> 1 pizza P grátis
// packages/shared/src/fidelidade.ts:1

export const META_FIDELIDADE = 10;
export const PREMIO_TAMANHO = "P" as const;

export type ClienteFidelidade = {
  telefone: string;
  pizzasGGanhas: number; //contador 0..9
  totalPizzasG: number;
  cuponsPendentes: number;
};

export function registrarVendaFidelidade(
  cliente: ClienteFidelidade,
  qtdPizzasG: number
): ClienteFidelidade & { novosCupons: number } {
  const total = cliente.pizzasGGanhas + qtdPizzasG;
  const novosCupons = Math.floor(total / META_FIDELIDADE);
  return {
    ...cliente,
    pizzasGGanhas: total % META_FIDELIDADE,
    totalPizzasG: cliente.totalPizzasG + qtdPizzasG,
    cuponsPendentes: cliente.cuponsPendentes + novosCupons,
    novosCupons,
  };
}

export function podeResgatar(cliente: ClienteFidelidade): boolean {
  return cliente.cuponsPendentes > 0;
}

export function resgatarCupom(cliente: ClienteFidelidade): ClienteFidelidade {
  if (!podeResgatar(cliente)) throw new Error("Sem cupons");
  return { ...cliente, cuponsPendentes: cliente.cuponsPendentes - 1 };
}

File: src/PizzaPDV.Core/Fidelidade.cs
namespace PizzaPDV.Core;

// Port de packages/shared/src/fidelidade.ts:1
// Regra: a cada 10 pizzas G -> 1 pizza P gratis
public static class Fidelidade
{
    public const int Meta = 10;
    public const string PremioTamanho = "P";

    public record ClienteFidelidade(string Telefone, int PizzasGGanhas, int TotalPizzasG, int CuponsPendentes);

    public static (ClienteFidelidade Cliente, int NovosCupons) RegistrarVenda(ClienteFidelidade cliente, int qtdPizzasG)
    {
        var total = cliente.PizzasGGanhas + qtdPizzasG;
        var novosCupons = total / Meta;
        var atualizado = cliente with
        {
            PizzasGGanhas = total % Meta,
            TotalPizzasG = cliente.TotalPizzasG + qtdPizzasG,
            CuponsPendentes = cliente.CuponsPendentes + novosCupons
        };
        return (atualizado, novosCupons);
    }

    public static bool PodeResgatar(ClienteFidelidade c) => c.CuponsPendentes > 0;

    public static ClienteFidelidade Resgatar(ClienteFidelidade c)
    {
        if (!PodeResgatar(c)) throw new InvalidOperationException("Sem cupons");
        return c with { CuponsPendentes = c.CuponsPendentes - 1 };
    }
}
